extension/spark-doris-connector/src/main/java/org/apache/doris/spark/backend/BackendClient.java

```diff
@@ -0,0 +1,218 @@
+                ex = e;
+            }
+            if (isConnected) {
+                logger.info("Success connect to {}.", routing);
+                break;
+            }
+        }
+        if (!isConnected) {
+            logger.error(ErrorMessages.CONNECT_FAILED_MESSAGE, routing);
+            throw new ConnectedFailedException(routing.toString(), ex);
+        }
+    }
+
+    private void close() {
+        logger.trace("Connect status before close with '{}' is '{}'.", routing, isConnected);
+        isConnected = false;
+        if (null != client) {
+            client = null;
+        }
+        if ((transport != null) && transport.isOpen()) {
+            transport.close();
+            logger.info("Closed a connection to {}.", routing);
+        }
+    }
+
+    /**
+     * Open a scanner for reading Doris data.
+     * @param openParams thrift struct to required by request
+     * @return scan open result
+     * @throws ConnectedFailedException throw if cannot connect to Doris BE
+     */
+    public TScanOpenResult openScanner(TScanOpenParams openParams) throws ConnectedFailedException {
+        logger.debug("OpenScanner to '{}', parameter is '{}'.", routing, openParams);
+        if (!isConnected) {
+            open();
+        }
+        TException ex = null;
+        for (int attempt = 0; attempt < retries; ++attempt) {
+            logger.debug("Attempt {} to openScanner {}.", attempt, routing);
+            try {
+                TScanOpenResult result = client.open_scanner(openParams);
+                if (result == null) {
+                    logger.warn("Open scanner result from {} is null.", routing);
+                    continue;
+                }
+                if (!TStatusCode.OK.equals(result.getStatus().getStatus_code())) {
+                    logger.warn("The status of open scanner result from {} is '{}', error message is: {}.",
+                            routing, result.getStatus().getStatus_code(), result.getStatus().getError_msgs());
+                    continue;
+                }
+                return result;
+            } catch (TException e) {
+                logger.warn("Open scanner from {} failed.", routing, e);
+                ex = e;
+            }
+        }
+        logger.error(ErrorMessages.CONNECT_FAILED_MESSAGE, routing);
+        throw new ConnectedFailedException(routing.toString(), ex);
+    }
+
+    /**
+     * get next row batch from Doris BE
+     * @param nextBatchParams thrift struct to required by request
+     * @return scan batch result
+     * @throws ConnectedFailedException throw if cannot connect to Doris BE
+     */
+    public TScanBatchResult getNext(TScanNextBatchParams nextBatchParams) throws ConnectedFailedException {
+        logger.debug("GetNext to '{}', parameter is '{}'.", routing, nextBatchParams);
+        if (!isConnected) {
+            open();
+        }
+        TException ex = null;
+        for (int attempt = 0; attempt < retries; ++attempt) {
```

Comment:
Maybe  getNext does not need to retry? Can you explain this?

Reply:
maybe failed because of network error.